[AI Chat] [Content Agent] Task UI

petemill opened the pull request:
Simplifies the display of regular tool use UI components, but adds a new Task UI that groups response events and allows the user to view either only the most recent LLM text completion and tool use requests, or a list of all the messages during the current event loop

<!-- Add brave-browser issue below that this PR will resolve -->
- Resolves https://github.com/brave/brave-browser/issues/49797

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

<!--
## Checklist:

- Review design docs
  [Browser design principles](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/chrome_browser_design_principles.md)
  [Style guide](https://chromium.googlesource.com/chromium/src/+/main/styleguide/c++/c++.md)
  [Core principles](https://www.chromium.org/developers/core-principles/)
- Ensure there are (tests)[https://www.chromium.org/developers/testing/]. Unit test as much as possible (including edge cases), but also include browser tests covering high level functionality.
- Ensure that there are comments explaining what classes/methods are/do. The "why" is often more important than the "what" in comments. Also update any relevant docs (moving docs from wiki to brave-core if necessary).
- Request security or other review (third-party libraries, rust code, etc...) if applicable [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) [other review](https://github.com/brave/reviews/issues/new/choose)
  Also see [adding third-party libraries](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/adding_to_third_party.md) for general guidelines on using third party code
- Maks sure there is a [ticket](https://github.com/brave/brave-browser/issues) for your issue
- Use Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- Write a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- Squash any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- Add appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- Run `git rebase master` (if needed)
-->

petemill commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/assistant_task.tsx:
```diff
@@ -0,0 +1,179 @@
+        <div className={styles.taskData}>
+          {showSteps && (
+            <Steps
+              {...props}
```
internal shared props

petemill commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/assistant_task.tsx:
```diff
@@ -4,13 +4,178 @@
+
+  return (
+    <div className={styles.task}>
+      <Tabs
```
Spacing is not quite right for inside the Tabs component and should probably be solved in `@brave/leo/Tabs`:
- https://github.com/brave/leo/issues/1238

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.ts:
```diff
@@ -0,0 +1,71 @@
+
+// TODO(https://github.com/brave/brave-browser/issues/48535):
+// Should be Navigate and TODO tools.
+const importantToolNames = [Mojom.NAVIGATE_TOOL_NAME]
```
Do you mind adding a comment explaining what makes a tool important & why we need to keep track of them?

petemill replied:
I have elaborated and made it clear it could change after collecting feedback:

```
// Important tools are ones worth always showing the latest version of
// to the user because they convey a longer-term status of the task compared
// to other tools. For example a navigation in a web page vs a click, or an
// up to date TODO list of items in the current task.
// TODO(https://github.com/brave/brave-browser/issues/48535):
// Should be Navigate and TODO tools.
// Note: Navigate could be removed if we display the current Url of a related
// content tab in the Task UI instead (since history navigation affects the
// "current" URL and not just the most recent navigation tool. This is all
// experimental.
```

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.ts:
```diff
@@ -0,0 +1,71 @@
+    )) {
+      if (!event) {
+        continue
+      }
```
can we actually have falsey events in the events array?

petemill replied:
I guess not. I thought something was complaining but will try

petemill replied:
removed in 246a466

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.ts:
```diff
@@ -0,0 +1,71 @@
+      if (!event) {
+        continue
+      }
+      if (!!event.completionEvent) {
```
mega nit - no need to coeerce as this is an if statement so we only care about the truthiness, we don't need it to be a boolean
```suggestion
      if (event.completionEvent) {
```

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.ts:
```diff
@@ -0,0 +1,71 @@
+        taskItems.push([event])
+      } else {
+        // Add any other event types to the last task item
+        taskItems[taskItems.length - 1].push(event)
```
```suggestion
        taskItems.at(-1).push(event)
```

petemill replied:
which I've added in 246a466

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.ts:
```diff
@@ -0,0 +1,71 @@
+    // Most recent of each type of important tool use event
+    const importantToolUseEvents: Partial<
+      Record<(typeof importantToolNames)[number], Mojom.ToolUseEvent>
+    > = {}
```
Hey does this work? `importantToolNames` isn't declared as `as const` so I supect this is a record of `<string, Mojom.ToolUseEvent>`

Probably we want an `as const` on the declaration

petemill replied:
The `as const` causes issues when using it for `importantToolNames.includes(event.toolUseEvent.toolName)`. You're right in that it's just treating it as a string, so I could just be explicit about that. It's not necessary to type the key, just to know it is a valid tool name, which is a string.

petemill replied:
changed in 246a466

bsclifton reviewed (approved):
string reviewers ++

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.ts:
```diff
@@ -0,0 +1,71 @@
+      taskItems,
+      allowedLinks,
+      importantToolUseEvents: Object.values(importantToolUseEvents).filter(
+        (event) => !!event,
```
Looking at the code above, we only insert non-falsey events, so do we need this?

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.ts:
```diff
@@ -0,0 +1,71 @@
+      }
+      if (event.completionEvent) {
+        // Start a new task item for every completion event
+        taskItems.push([event])
```
Hey sorry, does this mean completion events will be grouped with the next task?

`[1, 2, 'complete', 3, 4, 'complete']`
will result in
```
[1, 2]
['complete', 3, 4]
['complete']
```

Is that intentional? I (possibly naively) thought the completion should be for the tasks before:
```
[1, 2, 'complete']
[3, 4, 'complete']
```

I guess the above would be working if we were iterating in reverse order?

fallaciousreasoning replied:
That said, I'm pretty sure I'm just misunderstanding `completetionEvent`

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.test.ts:
```diff
@@ -0,0 +1,345 @@
+          (t) => t.toolName === Mojom.NAVIGATE_TOOL_NAME,
+        ),
+      ).toHaveLength(1)
+    })
```
I think we should just remove this until we get another tool

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.test.ts:
```diff
@@ -0,0 +1,345 @@
+      expect(result.current.importantToolUseEvents[0].id).toBe('3')
+    })
+
+    it('should not include non-important tools', () => {
```
can we have a test mixing important/non-important

petemill replied:
We do have that already with `'should extract an important tool'` or were you thinking to have some different kind of mix?

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.test.ts:
cool! Didn't think of testing hooks on their own like this but it makes a lot of sense!

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/assistant_task.module.scss:
```diff
@@ -0,0 +1,82 @@
+    top: calc(0.5lh - var(--half-task-icon-size));
+    left: calc(var(--half-task-icon-size) * -1);
+    width: calc(var(--task-icon-size) - calc(2 * var(--border-size)));
+    height: calc(var(--task-icon-size) - calc(2 * var(--border-size)));
```
just curious but any reason we can't use an outline here instead, so we don't need to worry about these size calculations?

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/assistant_task.stories.tsx:
```diff
@@ -5,16 +5,36 @@
 
 import * as React from 'react'
 import { Meta } from '@storybook/react'
+import { InferControlsFromArgs } from '../../../../../../.storybook/utils'
```
I guess not for this PR though :shrug:

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/assistant_task.tsx:
```diff
@@ -4,13 +4,177 @@
+
+  const currentCompletionEvent = lastTaskItem.find(
+    (event) => !!event.completionEvent,
+  )
```
At the moment, this can only ever be the first item in the taskItem - should we check that instead?

fallaciousreasoning reviewed (commented):
this looks great!

petemill commented on components/test/testPolyfills.ts:
```diff
@@ -21,3 +21,11 @@ const windowAsAny = window as any
 windowAsAny.chrome = getMockChrome()
 windowAsAny.loadTimeData = getMockLoadTimeData()
+
+windowAsAny.ResizeObserver = class ResizeObserverPolyfill {
```
Something already in `conversation_entries/index.test.ts` seems to rely on the nala code running. 

```
 ● ConversationEntries visualContentUsedPercentage handling › should show text content warning when visualContentUsedPercentage is fine but trimmed tokens exist

    expect(received).toBeInTheDocument()

    received value must be an HTMLElement or an SVGElement.
    Received has value: null

      473 |     // Should show text content warning due to trimmed tokens (highest priority)
      474 |     const infoElement = container.querySelector('leo-icon[name="info-outline"]')
    > 475 |     expect(infoElement).toBeInTheDocument()
```

petemill replied:
Seems fine to have the "polyfill" because the environment should at least know about that API anyway

petemill commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/assistant_task.tsx:
```diff
@@ -4,13 +4,177 @@
+      isRunnable
+      && !isActive
+      && taskItem.some(
+        (event) => !!event.toolUseEvent && !event.toolUseEvent.output,
```
done 246a466

petemill commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/assistant_task.tsx:
```diff
@@ -4,13 +4,177 @@
+  // "important" section.
+  const currentToolUseEvents = lastTaskItem.filter(
+    (event) =>
+      !!event.toolUseEvent
```
done

brave-builds commented:
> [!WARNING]
> You have got a presubmit warning. Please address it if possible.

```
Found 4 lines longer than 80 characters (first 5 shown).
```

Items:
```
components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/assistant_task.tsx, line 150, 85 chars
components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.test.ts, line 62, 85 chars
components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.ts, line 27, 81 chars
components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/use_extract_task_data.ts, line 52, 83 chars
```

<!-- presubmit-hash=61a516aa4411d036b62c3251164c262008d811670b326b912c4b8b929fdf6719 -->

petemill commented:
I rebased on master (new chromium update) and so invalidated the refs I added to review comments. But the feedback fixes are still split in to separate commits.

fallaciousreasoning commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_task/assistant_task.stories.tsx:
```diff
@@ -5,16 +5,36 @@
+        assistantEntries={taskConversationEntries}
+        isActiveTask={args.isActiveTask}
+        isGenerating={args.isGenerating}
+        isLeoModel={true}
```
mega nit:
```suggestion
        isLeoModel
```
